pkg/builder/builder.go

```diff
@@ -0,0 +1,213 @@
+package builder
+
+import (
+	"context"
+	"log"
+	"sync"
+
+	"github.com/go-logr/logr"
+	v1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	apiextv1 "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/typed/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/util/workqueue"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+func ControllerManagedBy(m manager.Manager) *Builder {
+	crdGetter, err := apiextv1.NewForConfig(m.GetConfig())
+	if err != nil {
+		log.Panic(err)
+	}
+
+	log := m.GetLogger().WithName("builder")
+
+	return &Builder{
+		blder:     ctrl.NewControllerManagedBy(m),
+		crdGetter: crdGetter,
+		log:       log,
+	}
+}
+
+// Builder builds a Controller and it can own objects later when the crd of the object created.
+type Builder struct {
+	blder *builder.Builder
+
+	ctrl             controller.Controller
+	crdGetter        apiextv1.CustomResourceDefinitionsGetter
+	forObject        runtime.Object
+	log              logr.Logger
+	globalPredicates []predicate.Predicate
+	tryOwnsInputs    []tryOwnsInput
+}
+
+func (blder *Builder) WithLogger(log logr.Logger) *Builder {
+	blder.blder.WithLogger(log)
+	blder.log = log
+
+	return blder
+}
+
+func (blder *Builder) For(object runtime.Object, opts ...builder.ForOption) *Builder {
+	blder.blder.For(object, opts...)
+	blder.forObject = object
+
+	return blder
+}
+
+func (blder *Builder) Owns(object runtime.Object, opts ...builder.OwnsOption) *Builder {
+	blder.blder.Owns(object, opts...)
+
+	return blder
+}
+
+func (blder *Builder) WithEventFilter(p predicate.Predicate) *Builder {
+	blder.blder.WithEventFilter(p)
+	blder.globalPredicates = append(blder.globalPredicates, p)
+
+	return blder
+}
+
+func (blder *Builder) WithOptions(options controller.Options) *Builder {
+	blder.blder.WithOptions(options)
+
+	return blder
+}
+
+// TryOwns owns the object when the crdDependency found, otherwise try owns it after the crdDependency found.
+func (blder *Builder) TryOwns(object runtime.Object, crdDependency string, predicates ...predicate.Predicate) *Builder {
```

Comment:
Suggest changing the name to TryOwnsCRD, because it only works for CRD

Reply:
The best name for this method is `TryOwnsWhenCRDReady` but this is too long.